communication/src/AsyncTlsConnection.cpp

```diff
@@ -185,82 +170,97 @@ uint32_t AsyncTlsConnection::getReadMsgSize() {
   return ntohl(num);
 }
 
+void AsyncTlsConnection::remoteDispose() {
+  auto self = shared_from_this();
+  asio::post(strand_, [this, self] {
+    static constexpr bool close_connection = false;
+    dispose(close_connection);
+  });
+}
+
 void AsyncTlsConnection::dispose(bool close_connection) {
   // We only want to dispose of a connection if it was actually authenticated, in which case it has
   // started to be used. In the case of a server connection failing authentication, it will have no
   // corresponding peer_id_.
   if (disposed_ || !peer_id_.has_value()) return;
   LOG_WARN(logger_, "Closing connection to node " << peer_id_.value());
   disposed_ = true;
-  boost::system::error_code _;
-  read_timer_.cancel(_);
-  write_timer_.cancel(_);
+  read_timer_.cancel();
+  write_timer_.cancel();
+  auto self = shared_from_this();
   if (close_connection) {
-    tlsTcpImpl_.closeConnection(peer_id_.value());
+    // The ConnMgr runs in a separate strand. We must post a message to inform it, rather than calling directly.
+    conn_mgr_.remoteCloseConnection(peer_id_.value());
   }
 }
 
-void AsyncTlsConnection::write() {
-  if (disposed_ || write_msg_) return;
+void AsyncTlsConnection::send(std::shared_ptr<OutgoingMsg>&& msg) {
+  auto self = shared_from_this();
+  asio::post(strand_, [this, self, msg{move(msg)}]() { write(msg); });
+}
 
-  write_msg_ = write_queue_->pop();
-  if (!write_msg_) return;
+void AsyncTlsConnection::write(std::shared_ptr<OutgoingMsg> msg) {
+  if (disposed_ || !msg) return;
```

Comment:
It can be NULL. The call at line 240 can return a null pointer if the write queue is empty. It was a tradeoff vs adding 2 more lines before the recursive call.

Reply:
Thank you for the explanation!